File: packages/typescript/src/result.ts
import type { CostEstimate } from './cost';

/**
 * Result of a SQL validation check.
 *
 * Can be used in boolean context:
 * ```typescript
 * const result = validate("SELECT * FROM users");
 * if (result) {
 *   // Query is safe
 * }
 * ```
 */
export class ValidationResult {
  /** Whether the query passed all validation checks */
  readonly isSafe: boolean;

  /** Explanation if the query was blocked (undefined if safe) */
  readonly reason?: string;

  /** The type of SQL statement (SELECT, INSERT, DROP, etc.) */
  readonly statementType?: string;

  /** Tables referenced in the query */
  readonly tables: string[];

  /** Estimated query cost/complexity (undefined if cost estimation disabled) */
  readonly cost?: CostEstimate;

  /** The LIMIT value if present in the query */
  readonly limitValue?: number;

  /** Non-blocking issues detected (e.g., high cost queries that passed) */
  readonly warnings: string[];

  private constructor(data: {
    isSafe: boolean;
    reason?: string;
    statementType?: string;
    tables?: string[];
    cost?: CostEstimate;
    limitValue?: number;
    warnings?: string[];
  }) {
    this.isSafe = data.isSafe;
    this.reason = data.reason;
    this.statementType = data.statementType;
    this.tables = data.tables ?? [];
    this.cost = data.cost;
    this.limitValue = data.limitValue;
    this.warnings = data.warnings ?? [];
  }

  /**
   * Create a safe validation result.
   */
  static safe(
    meta: {
      statementType?: string;
      tables?: string[];
      cost?: CostEstimate;
      limitValue?: number;
      warnings?: string[];
    } = {}
  ): ValidationResult {
    return new ValidationResult({ isSafe: true, ...meta });
  }

  /**
   * Create an unsafe validation result with a reason.
   */
  static unsafe(
    reason: string,
    meta: {
      statementType?: string;
      tables?: string[];
      cost?: CostEstimate;
      limitValue?: number;
      warnings?: string[];
    } = {}
  ): ValidationResult {
    return new ValidationResult({ isSafe: false, reason, ...meta });
  }

  /**
   * Allow using ValidationResult in boolean context.
   * Note: This is for documentation purposes; JS doesn't support operator overloading.
   * Use result.isSafe directly.
   */
  valueOf(): boolean {
    return this.isSafe;
  }

  toString(): string {
    if (this.isSafe) {
      return `ValidationResult(safe, type=${this.statementType ?? 'unknown'})`;
    }
    return `ValidationResult(unsafe, reason="${this.reason}")`;
  }
}

File: packages/python/proxql/result.py
"""Validation result types."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .cost import CostEstimate


@dataclass(frozen=True)
class ValidationResult:
    """Immutable result of SQL validation.

    Attributes:
        is_safe: Whether the query passed all validation checks.
        reason: Human-readable explanation if blocked (None if safe).
        statement_type: The type of SQL statement (SELECT, INSERT, DROP, etc.).
        tables: List of table names referenced in the query.
        cost: Estimated query cost/complexity (None if cost estimation disabled).
        limit_value: The LIMIT value if present in the query.
        warnings: Non-blocking issues detected (e.g., high cost queries that passed).
    """

    is_safe: bool
    reason: str | None = None
    statement_type: str | None = None
    tables: list[str] = field(default_factory=list)
    cost: CostEstimate | None = None
    limit_value: int | None = None
    warnings: list[str] = field(default_factory=list)

    def __bool__(self) -> bool:
        """Allow using result directly in boolean context."""
        return self.is_safe
